### PR TITLE
mgr/dashboard: Automate system user creation and restart of rgw daemons on Migrate to Multisite

### PR DESCRIPTION
**Pull request by aaSharma14**
1. Automate the process of replication user creation during single site to multi-site migration
2. Provide a support for automatic restart post creation of default realm/zone group/zone as a part of multisite migrate

Fixes: https://tracker.ceph.com/issues/69052

Signed-off-by: Aashish Sharma <aasharma@redhat.com>

https://github.com/user-attachments/assets/be7b9256-db32-402a-b830-2541cda73083

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-migrate/rgw-multisite-migrate.component.html [hunk 1]
```diff
@@ -111,36 +111,18 @@
...
       <div class="form-group row">
         <label class="cd-col-form-label required"
-               for="access_key"
-               i18n>S3 access key
-          <cd-helper>
-            <span>To see or copy your S3 access key, go to <b>Object Gateway > Users</b> and click on your user name. In <b>Keys</b>, click <b>Show</b>. View the access key by clicking Show and copy the key by clicking <b>Copy to Clipboard</b>.</span>
-          </cd-helper>
-        </label>
+               for="username"
+               i18n>Username</label>
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-migrate/rgw-multisite-migrate.component.html:115]: I'd recommend to add a helper here stating that this is to add a new user and not adding an existing to user of rgw.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-migrate/rgw-multisite-migrate.component.ts [hunk 2]
```diff
@@ -174,11 +175,10 @@ export class RgwMultisiteMigrateComponent implements OnInit {
     this.zone = new RgwZone();
     this.zone.name = values['zoneName'];
     this.zone.endpoints = values['zone_endpoints'];
-    this.zone.system_key = new SystemKey();
-    this.zone.system_key.access_key = values['access_key'];
-    this.zone.system_key.secret_key = values['secret_key'];
-    this.rgwMultisiteService.migrate(this.realm, this.zonegroup, this.zone).subscribe(
+    const username = values['username'];
+    this.rgwMultisiteService.migrate(this.realm, this.zonegroup, this.zone, username).subscribe(
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-migrate/rgw-multisite-migrate.component.ts:179]: if `username` has no other purpose, let's use that directly in the function.
```suggestion
    this.rgwMultisiteService.migrate(this.realm, this.zonegroup, this.zone, values['username']).subscribe(
```

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 3]
```diff
@@ -1411,19 +1410,34 @@ def migrate_to_multisite(self, realm_name: str, zonegroup_name: str, zone_name:
...
+                                    zonegroup_name=zonegroup_name,
+                                    default='true', master='true',
+                                    endpoints=zone_endpoints,
+                                    access_key=keys['access_key'],
+                                    secret_key=keys['secret_key'])
+                else:
+                    raise ValueError("Access key or secret key is missing")
+        except Exception as e:
+            logger.error("Failed to modify zone or create system user: %s", e)
+            raise
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/services/rgw_client.py:1432]: shouldn't we raise a Dashboard Exception stating the error?

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 4]
```diff
@@ -1411,19 +1410,34 @@ def migrate_to_multisite(self, realm_name: str, zonegroup_name: str, zone_name:
...
+
+        try:
+            rgw_service_manager = RgwServiceManager()
+            rgw_service_manager.restart_rgw_daemons_and_set_credentials()
+        except Exception as e:
+            logger.error("Failed to restart RGW daemon: %s", e)
+            raise
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/services/rgw_client.py:1439]: same as above

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 5]
```diff
@@ -1411,19 +1410,34 @@ def migrate_to_multisite(self, realm_name: str, zonegroup_name: str, zone_name:
...
+                secret_key = keys['secret_key']
+                if access_key and secret_key:
+                    self.modify_zone(zone_name=zone_name,
+                                    zonegroup_name=zonegroup_name,
+                                    default='true', master='true',
+                                    endpoints=zone_endpoints,
+                                    access_key=keys['access_key'],
+                                    secret_key=keys['secret_key'])
+                else:
+                    raise ValueError("Access key or secret key is missing")
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/services/rgw_client.py:1429]: why are we raising a ValueError? this should also raise a DashboardException so the UI can properly catch it.

**Review by nizamial09 — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-migrate/rgw-multisite-migrate.component.html (hunk 1 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-migrate/rgw-multisite-migrate.component.html:115]: Thanks @nizamial09 , Updated

**Review by nizamial09 — APPROVED** [review]
code looks good. thank you @aaSharma14

**Comment by aaSharma14** [comment]
jenkins test api

**Comment by aaSharma14** [comment]
jenkins test api

**Review by nmunet — APPROVED** [review]
Tested, LGTM

**Pull request merged**